### ISSUE TITLE
DESKTOP: Paste does not work for images

### ISSUE THREAD

**Issue by Rishgod** [laurent22/joplin#2663]
## Description
Paste function for Images do not work the way it should me (both Right click-> paste and Ctrl+V) when pasting images to Joplin editor.

## Actual behaviour
![paste-bug](https://user-images.githubusercontent.com/54576074/75954987-f8e72200-5eda-11ea-83cd-ddb0a475cb02.gif)
## Expected Behaviour
![paste-expected-behaviour](https://user-images.githubusercontent.com/54576074/75955845-ad357800-5edc-11ea-87c3-6d0e7a53a32a.gif)

### Tasks
- [ ] Enable Paste feature for images
- [ ] Ensure it works for both the methods (Right click-> paste and Ctrl+V) 

@PackElend  @tessus @laurent22  @bedwardly-down what do you guys think about this, and if this behaviour is indeed unexpected and should be fixed then **_I would like to work on it_**.

**Comment by laurent22** [comment]
Thanks for looking into it @Rishgod, it's indeed a bug that would need to be fixed. If you can look at it that would be great.

**Comment by bedwardly-down** [comment]
There's also another three other things that happen (at least on mobile): 
1. pasting any images with text in them, Joplin turns the text part into a text entry in your note instead of pasting the image (could be OCR related)
2. attaching images using the `Attach image` function doesn't place the image at the cursor but at the bottom of the not 
 3. when multiple images are attached at one time, they aren't attached in proper order but seemingly random placement 

Since all clients share the same code for much of the core functions, is it possible these could all be caused by the same bug in the code or be separate bugs altogether?

**Comment by laurent22** [comment]
> pasting any images with text in them, Joplin turns the text part into a text entry in your note instead of pasting the image (could be OCR related)

It's been looked at by several people and so far no good fix for this. When there's text and image at the same time, it's hard to guess what should be pasted (there are several discussions about this here and the forum).

> when multiple images are attached at one time, they aren't attached in proper order but seemingly random placement

Perhaps add a bug about this? I guess they should be sorted in alphabetical order? I found that this issue is not really Joplin specific. Whenever I import multiple files in any app, the order is often random.

> attaching images using the Attach image function doesn't place the image at the cursor but at the bottom of the not

It indeed should insert at cursor so feel free to create a separate GH issue for this.

**Comment by bedwardly-down** [comment]
Heading to work but will try to do here in a bit. Thanks

**Comment by deadpool1999** [comment]
Copying an image from a web browser (chrome ) and then doing the above mentioned operations{ctrl+v} gives the desired result!!
It's some security issue with chrome i guess, like for a div element with contentEditable="true" everything works fine as long as the copy paste happens b/w similar environments.
(So copying from chrome and pasting the image in my note works fine, but the same isn't true for a remote location(desktop).  )

**Comment by Rishgod** [comment]
> Copying an image from a web browser (chrome ) and then doing the above mentioned operations{ctrl+v} gives the desired result!!
> It's some security issue with chrome i guess, like for a div element with contentEditable="true" everything works fine as long as the copy paste happens b/w similar environments.
> (So copying from chrome and pasting the image in my note works fine, but the same isn't true for a remote location(desktop). )

@deadpool1999  thanks for pointing it out, I look into the same.

**Comment by yasithmilinda** [comment]
Hello everyone, would you mind if I take on this issue?

**Comment by Rishgod** [comment]
> Hello everyone, would you mind if I take on this issue?

hey @yasithmilinda I' almost done with this, mind taking some other issue :)

**Comment by yasithmilinda** [comment]
Hey @Rishgod, no problem! :) I'll look into other issues, thanks!

**Comment by Rishgod** [comment]
After doing a lot of research and trying everything it is **not possible to Implement this feature**



<img width="373" alt="B3" src="https://user-images.githubusercontent.com/54576074/76152271-9ca31e80-60e3-11ea-892b-fa3a703d76d6.png">



When I was trying to fix this, I was really surprised to find out that `clipboardImageCheck` is always true, irrespective of the fact that the clipboard has an image or not.



![b0-gif](https://user-images.githubusercontent.com/54576074/76152276-af1d5800-60e3-11ea-8394-eda7ab2e4fe5.gif)


<hr/>


**When you copy an image in Windows Explorer, you are copying the file instead of copying the contents of the image.** But that's, not the case if you take a screenshot or copy the photo from a picture app.

![b2-gif](https://user-images.githubusercontent.com/54576074/76152275-ab89d100-60e3-11ea-8a76-ae092f1f519e.gif)



It does not look like Electron supports this feature, I google a lot about it and I also found a lot of [similar issue](https://github.com/atom/tree-view/issues/569) on GitHub.


There are certain workaround to it I found on Github, [here](https://github.com/electron/electron/issues/9035) 

It basically how to get file pointers in the clipboard, but it can only get only one file's path if multi files are copied(the very first image if you select many).


@laurent22 Thoughts?

**Comment by bedwardly-down** [comment]
@Rishgod, i found [this](https://code-examples.net/en/q/5dd891) Any way this could be used to basically take the clipboard, convert it to base64 in react (since electron doesn't support it but there may be a way to access the clipboard that way, still), and then rebuild the image from that data into a format than be displayed in app?

**Comment by deadpool1999** [comment]
Yes , electron js uses node and chromium , so this issue can't be solved. This issue also exists with the word online and also the docs online version (when opened in chrome) but it is not the same if opened in Firefox ! Trying to read the clipboard content also shows 0 items in it. So can't also use some other api to upload the image somewhere and then fetch it in chrome!

**Comment by tessus** [comment]
@PackElend we have an `upstream` label. Does your `external` mean the same?

**Comment by PackElend** [comment]
it properly means the same; i refer to code outside the Joplin universe, so basically 3rd party dependencies  

I renamed the label to allow better filtering on externals.

**Comment by ffgm** [comment]
@laurent22

**Comment by kuraci** [comment]
srtongly hope this bug fixed.
when copy only one image to joplin，it copied image， no problem.
but if copy more, such as with some words , it copied only words. image would lost while link of image left. image would display well in loacal device but remote device go missing the image.

**Comment by madrix01** [comment]
Hello everyone, I'm an intermediate developer, I love open source and looking forward for participating in GSoC'22. I want to start contributing to Joplin. Can someone guide me for the same and I also wanted to know how to contact a mentor for GSoC'22.
Thank you!

**Comment by laurent22** [comment]
There's no mentor to contact yet since we have nothing ready for 2022, but if in the meantime you want to get used to the codebase, a good option is to check the issue trackers from time to time and see if you can work on the issues with labels "bug" and "high". Or alternatively you can try to find some good plugin idea and work on it.

**Comment by robertchen** [comment]
copy HTML and paste to word, word has images, but copy html contents paste to joplin, the images are lost

**Comment by oryoruk** [comment]
I am having this issue (Joplin 2.14.22). I can drag and drop images fine, but cannot paste images/image screenshots into the markdown or rich-text editors.

**Comment by srajan0149** [comment]
I think this issue has been resolved, since it works on my system. Can anybody else confirm?
